### ISSUE TITLE
Some Python blocks may be able to be moved into C++

### ISSUE THREAD

**Issue by ncorgan** [gnuradio/gnuradio#4944]
There are a bunch of blocks (mostly hierarchical) developed in Python, I assume out of ease of development. This stops them from being used by C++ clients (or other hierarchical blocks). In theory, some of the could be reimplemented on the C++ layer. With GNU Radio's Python bindings (and maybe an interim compatibility `.py` file if necessary), these shouldn't break any Python clients, and the QA tests should still work.

Some of them use Python-specific stuff (socket stuff comes to mind) but certainly not all. These days, y'all seem to be more open to change than in the past, so I thought I'd bring it up.

**Comment by mormj** [comment]
Can you start a checklist here of blocks that would be good candidates?

**Comment by ncorgan** [comment]
I've gone through the first few modules and listed the technically viable blocks. I'll continue tomorrow.

**Comment by ncorgan** [comment]
@mormj Done

**Comment by solomonbstoner** [comment]
@ncorgan is the suggestion to create hier blocks in c++ to replace these hier blocks currently in python?

**Comment by willcode** [comment]
Not totally related to this issue, but throwing it out there ...

When a block is used in a hier block, there will be only one downstream block. In-place operations are possible in many cases then, if blocks are written to support this. Separating the work "kernel" of a block from the `work()` of a block would allow multiple functions to be called from a hier block without the need to more threads, buffers, latency. This would not require (any? much?) scheduler work, just some changes to hier and participating blocks, and it would all be backward compatible.

**Comment by ncorgan** [comment]
@solomonbstoner Correct. It should be possible to do this without breaking Python clients.

@willcode Ah, I wasn't aware of any non-block changes that would be needed. I assumed it would be a 1-to-1 conversion to some extent. That's obviously something to take into account for the necessary effort. FWIW, this doesn't apply to all Python conversions proposed above.

**Comment by willcode** [comment]
@ncorgan I don't think there's any non-block work required for your idea (need to read all the blocks to know for sure). My comment was additional idea.

**Comment by willcode** [comment]
@ncorgan Yah, there is a bunch of util stuff in python that would take some effort to port, and python versions would have to remain for backward compatibility. Could start with the easy/most valuable ones.

**Comment by drmpeg** [comment]
For gr-dtv, those Python blocks could just be deleted. The ATSC receiver was completely reworked last year and doesn't use ```atsc_filter.py``` anymore. ```atsc_filter.py and ```atsc_rx.py``` are only used by ```file_atsc_rx.grc``` which could easily be deleted since it does the same thing as uhd_atsc_rx.grc (if you just add a file source block).

**Comment by marcusmueller** [comment]
@drmpeg I'm hesitant to delete stuff abruptly, but would marking them "deprecated" on maint-3.9 and deleting them on master sound like a good idea?

**Comment by drmpeg** [comment]
@marcusmueller Sure, I'm fine with that.

**Comment by ncorgan** [comment]
WIP porting optfir (arbitrarily chosen): https://github.com/ncorgan/gnuradio/tree/optfir-to-cpp

**Comment by willcode** [comment]
Backport note: `for_each` is C++-17.

**Comment by ncorgan** [comment]
Ah, force of habit. Let me know if I need to defancy it, probably in the PR.

**Comment by marcusmueller** [comment]
I, for one, think backporting might be a bit risky here.

**Comment by willcode** [comment]
Definitely more work to backport. Up to @ncorgan whether it's worth it.
